File: project/src/nwr/gui/TeachWindow.java
/*
 *  "NorseWorld: Ragnarok", a roguelike game for PCs.
 *  Copyright (C) 2002-2008, 2014 by Serg V. Zhdanovskih (aka Alchemist).
 *
 *  This file is part of "NorseWorld: Ragnarok".
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package nwr.gui;

import jzrlib.core.Rect;
import nwr.creatures.NWCreature;
import nwr.core.Locale;
import nwr.core.StaticData;
import nwr.core.types.AbilityID;
import nwr.core.types.SkillID;
import nwr.core.types.TeachableKind;
import nwr.engine.BaseControl;
import nwr.engine.BaseScreen;
import nwr.engine.MouseButton;
import nwr.gui.controls.LBItem;
import nwr.gui.controls.LBOptions;
import nwr.gui.controls.ListBox;
import nwr.main.GlobalVars;
import nwr.player.Player;

/**
 *
 * @author Serg V. Zhdanovskih
 */
public final class TeachWindow extends DialogWindow
{
    private static final class TeachItem
    {
        public int Index;
        public int CurLev;
        public int Price;
        
        public TeachItem(int index, int curLev, int price)
        {
            this.Index = index;
            this.CurLev = curLev;
            this.Price = price;
        }
    }
    
    private final ListBox fDisciplinesList;

    public TeachWindow(BaseControl owner)
    {
        super(owner);
        this.fDisciplinesList = new ListBox(this);
        this.fDisciplinesList.setMode(ListBox.MODE_REPORT);
        this.fDisciplinesList.setColumns(2);

        this.fDisciplinesList.setBounds(new Rect(10, 10, 589, 398));

        this.fDisciplinesList.setItemHeight(34);
        this.fDisciplinesList.setIconHeight(30);
        this.fDisciplinesList.setIconWidth(32);
        this.fDisciplinesList.Options.include(LBOptions.lboIcons);
        this.fDisciplinesList.getColumnTitles().add("", 460);
        this.fDisciplinesList.getColumnTitles().add("", 72);
        this.fDisciplinesList.setVisible(true);
        this.fDisciplinesList.setColumns(1);
        this.fDisciplinesList.OnItemSelect = this::onDisciplineSelect;
    }

    @Override
    protected void dispose(boolean disposing)
    {
        if (disposing) {
            this.fDisciplinesList.dispose();
        }
        super.dispose(disposing);
    }

    @Override
    protected void doShowEvent()
    {
        super.doShowEvent();
        this.fDisciplinesList.ImagesList = GlobalVars.nwrWin.Resources;
        this.fDisciplinesList.setIconHeight(30);
        this.fDisciplinesList.setIconWidth(32);
        this.updateView();
    }

    private void onDisciplineSelect(Object sender, MouseButton button, LBItem item)
    {
        if (button == MouseButton.mbRight) {
            TeachItem ti = (TeachItem) item.Data;
            if (ti.Price <= GlobalVars.nwrGame.getPlayer().getMoney()) {
                GlobalVars.nwrGame.getPlayer().teachDiscipline(this.fCollocutor, ti.Index, ti.CurLev);
                this.updateView();
            }
        }
    }

    private void updateView()
    {
        this.fDisciplinesList.getItems().beginUpdate();
        this.fDisciplinesList.getItems().clear();
        NWCreature clt = this.fCollocutor;
        Player p = GlobalVars.nwrGame.getPlayer();

        for (int i = 0; i < StaticData.dbTeachable.length; i++) {
            int id = StaticData.dbTeachable[i].id;
            boolean res = false;
            TeachableKind kind = StaticData.dbTeachable[i].kind;

            String s = "";
            int imageIndex = -1;
            int curLev = 0;
            switch (kind) {
                case Ability:
                    AbilityID ab = AbilityID.forValue(id);
                    if (clt.getAbility(ab) > 0) {
                        s = Locale.getStr(ab.NameRS);
                        imageIndex = -1;
                        curLev = p.getAbility(ab);
                        res = (curLev < clt.getAbility(ab));
                    }
                    break;

                case Skill:
                    SkillID sk = SkillID.forValue(id);
                    if (clt.getSkill(sk) >= 0) {
                        s = Locale.getStr(sk.NameRS);
                        imageIndex = sk.ImageIndex;
                        curLev = p.getSkill(sk);
                        res = (curLev < clt.getSkill(sk));
                    }
                    break;
            }

            int price = (int) GlobalVars.nwrGame.getTeachablePrice(i, curLev);

            if (res) {
                String st = " ( " + String.valueOf(curLev) + " -> " + String.valueOf(curLev + 1) + " )";
                LBItem listItem = this.fDisciplinesList.getItems().add(s + st, new TeachItem(i, curLev, price));
                if (price > p.getMoney()) {
                    listItem.Color = BaseScreen.clRed;
                } else {
                    listItem.Color = BaseScreen.clGold;
                }
                listItem.ImageIndex = imageIndex;
                listItem.getSubItems().add(String.valueOf(price), null);
            }
        }

        this.fDisciplinesList.getItems().endUpdate();
    }
}

File: project/GUI/TeachWindow.cs
/*
 *  "NorseWorld: Ragnarok", a roguelike game for PCs.
 *  Copyright (C) 2002-2008, 2014 by Serg V. Zhdanovskih.
 *
 *  This file is part of "NorseWorld: Ragnarok".
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

using System;
using BSLib;
using NWR.Creatures;
using NWR.Game;
using NWR.Game.Types;
using NWR.GUI.Controls;
using ZRLib.Core;
using ZRLib.Engine;

namespace NWR.GUI
{
    public sealed class TeachWindow : DialogWindow
    {
        private sealed class TeachItem
        {
            public int Index;
            public int CurLev;
            public int Price;

            public TeachItem(int index, int curLev, int price)
            {
                Index = index;
                CurLev = curLev;
                Price = price;
            }
        }

        private readonly ListBox fDisciplinesList;

        public TeachWindow(BaseControl owner)
            : base(owner)
        {
            fDisciplinesList = new ListBox(this);
            fDisciplinesList.Mode = ListBox.MODE_REPORT;
            fDisciplinesList.Columns = 2;

            fDisciplinesList.Bounds = ExtRect.Create(10, 10, 589, 398);

            fDisciplinesList.ItemHeight = 34;
            fDisciplinesList.IconHeight = 30;
            fDisciplinesList.IconWidth = 32;
            fDisciplinesList.Options.Include(LBOptions.lboIcons);
            fDisciplinesList.ColumnTitles.Add("", 460);
            fDisciplinesList.ColumnTitles.Add("", 72);
            fDisciplinesList.Visible = true;
            fDisciplinesList.Columns = 1;
            fDisciplinesList.OnItemSelect = OnDisciplineSelect;
        }

        protected override void Dispose(bool disposing)
        {
            if (disposing) {
                fDisciplinesList.Dispose();
            }
            base.Dispose(disposing);
        }

        protected override void DoShowEvent()
        {
            base.DoShowEvent();
            fDisciplinesList.ImagesList = GlobalVars.nwrWin.Resources;
            fDisciplinesList.IconHeight = 30;
            fDisciplinesList.IconWidth = 32;
            UpdateView();
        }

        private void OnDisciplineSelect(object sender, MouseButton button, LBItem item)
        {
            if (button == MouseButton.mbRight) {
                TeachItem ti = (TeachItem)item.Data;
                if (ti.Price <= GlobalVars.nwrGame.Player.Money) {
                    GlobalVars.nwrGame.Player.TeachDiscipline(fCollocutor, ti.Index, ti.CurLev);
                    UpdateView();
                }
            }
        }

        private void UpdateView()
        {
            fDisciplinesList.Items.BeginUpdate();
            fDisciplinesList.Items.Clear();
            NWCreature clt = fCollocutor;
            Player p = GlobalVars.nwrGame.Player;

            for (int i = 0; i < StaticData.dbTeachable.Length; i++) {
                int id = StaticData.dbTeachable[i].Id;
                bool res = false;
                TeachableKind kind = StaticData.dbTeachable[i].Kind;

                string s = "";
                int imageIndex = -1;
                int curLev = 0;
                switch (kind) {
                    case TeachableKind.Ability:
                        AbilityRec abRec = StaticData.dbAbilities[id];
                        AbilityID ab = (AbilityID)id;
                        if (clt.GetAbility(ab) > 0) {
                            s = BaseLocale.GetStr(abRec.Name);
                            imageIndex = -1;
                            curLev = p.GetAbility(ab);
                            res = (curLev < clt.GetAbility(ab));
                        }
                        break;

                    case TeachableKind.Skill:
                        SkillRec skRec = StaticData.dbSkills[id];
                        SkillID sk = (SkillID)id;
                        if (clt.GetSkill(sk) >= 0) {
                            s = BaseLocale.GetStr(skRec.Name);
                            imageIndex = skRec.ImageIndex;
                            curLev = p.GetSkill(sk);
                            res = (curLev < clt.GetSkill(sk));
                        }
                        break;
                }

                int price = (int)GlobalVars.nwrGame.GetTeachablePrice(i, curLev);

                if (res) {
                    string st = " ( " + Convert.ToString(curLev) + " -> " + Convert.ToString(curLev + 1) + " )";
                    LBItem listItem = fDisciplinesList.Items.Add(s + st, new TeachItem(i, curLev, price));
                    if (price > p.Money) {
                        listItem.Color = Colors.Red;
                    } else {
                        listItem.Color = Colors.Gold;
                    }
                    listItem.ImageIndex = imageIndex;
                    listItem.SubItems.Add(Convert.ToString(price), null);
                }
            }

            fDisciplinesList.Items.EndUpdate();
        }
    }

}